hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java

```diff
@@ -18,27 +18,139 @@
+        final byte[] pseudoRandomBytes = getNMinus3RunningHashBytes(nMinusThreeRunningHash);
+
+        // If no bytes are available then return
+        if (pseudoRandomBytes == null || pseudoRandomBytes.length == 0) {
+            return;
+        }
+        // If range is provided then generate a random number in the given range
+        // from the pseudoRandomBytes
+        if (range > 0) {
+            final int pseudoRandomNumber = randomNumFromBytes(pseudoRandomBytes, range);
+            recordBuilder.setPrngNumber(pseudoRandomNumber);
+        } else {
+            recordBuilder.setPrngBytes(Bytes.wrap(pseudoRandomBytes));
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PrngRecordBuilder newRecordBuilder() {
+        return new UtilPrngRecordBuilder();
+    }
+
+    /**
+     * Get the n-3 running hash bytes from the n-3 running hash. If the running hash is not
+     * available, will throw a HandleException. n-3 running hash is chosen for generating random
+     * number instead of n-1 running hash for processing transactions quickly. Because n-1 running
+     * hash might not be available when the transaction is processed and might need to wait longer.
+     *
+     * @param nMinus3RunningHash n-3 running hash
+     * @return n-3 running hash bytes
+     */
+    private byte[] getNMinus3RunningHashBytes(@NonNull final RunningHash nMinus3RunningHash) {
+        // This can't happen because this running hash is taken from record stream.
+        // If this happens then there is a bug in the code.
+        requireNonNull(nMinus3RunningHash);
+
+        Hash nMinusThreeHash;
+        try {
+            nMinusThreeHash = nMinus3RunningHash.getFutureHash().get();
+            // Use n-3 running hash instead of n-1 running hash for processing transactions quickly
+            if (nMinusThreeHash == null || Arrays.equals(nMinusThreeHash.getValue(), new byte[48])) {
+                log.info("No n-3 record running hash available to generate random number");
+                return MISSING_BYTES;
+            }
+            // generate binary string from the running hash of records
+            return nMinusThreeHash.getValue();
+        } catch (InterruptedException | ExecutionException e) {
+            log.error("Interrupted exception while waiting for n-3 running hash", e);
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException(e);
+            // TODO: Handle exception to be thrown here. Need to decide on the status to be
+            //  thrown for any interrupted exception
+        }
+    }
+
+    /**
+     * Generate a random number from the given bytes in the given range
+     * @param pseudoRandomBytes bytes to generate random number from
+     * @param range range of the random number
+     * @return random number
+     */
+    private int randomNumFromBytes(@NonNull final byte[] pseudoRandomBytes, final int range) {
```

Comment:
Not sure how crucial real randomness is in this context, but I think depending on the `range` this function will return slightly skewed results. Is this acceptable?

Reply:
Yeah, since we are not changing the business logic and it is how the HIP specified. I think it is fine